### DISCUSSION TITLE
toLocaleString for PlainMonthDay and PlainYearMonth throws

### DISCUSSION THREAD

**Discussion by mrcljx** [js-temporal/temporal-polyfill#182] [Q&A] [answered]
I was surprised to find that `toLocaleString` does not work by default on a lot of the plain types, because it refuses to format iso8601 with the default (gregorian) calendar.

```
> Temporal.PlainMonthDay.from("04-01").toLocaleString()
Uncaught:
RangeError: cannot format PlainMonthDay with calendar iso8601 in locale with calendar gregory
```

```
> Temporal.PlainYearMonth.from("2022-01").toLocaleString()
Uncaught:
RangeError: cannot format PlainYearMonth with calendar iso8601 in locale with calendar gregory
```

while `PlainDate` has an exception for `iso8601`.

https://github.com/js-temporal/temporal-polyfill/blob/main/lib/intl.ts#L584

```
> Temporal.PlainDate.from("2022-04-01").toLocaleString()
'4/1/2022'
```

This seems inconsistent to me. Can someone point me to where this is mandated by the spec, or is that something this implementation could change?

**Comment by ptomato** [comment, marked as answer]
This is by design — the reason it's inconsistent is because you can convert a PlainDate from one calendar system to another, but not a PlainYearMonth or PlainMonthDay. At least, not without additional information (like, supplying the missing day-of-month for PlainYearMonth or year for PlainMonthDay.) It's true that "iso8601" to "gregory" is one of the few cases where you can just map the ISO reference date onto the Gregorian reference date and it will be correct. But if we allowed that for certain calendar pairs, we'd have the problem that toLocaleString() might work for a developer converting ISO to Gregorian, but throw for a user whose locale had a different calendar selected.

The simplest way to get localized formatting for PlainYearMonth and PlainMonthDay is to pass the calendar in the toLocaleString options. For example:

```js
plainYearMonth.toLocaleString(locale, { calendar: plainYearMonth.calendar })
```

Another way that is more complicated, but might be what you want in your application, is to supply the missing information for the conversion from one calendar to the other, in a way that works for your application. For example, for a PlainYearMonth you might decide that you want the PlainYearMonth in the locale's calendar that day 1 of the original PlainYearMonth falls on:

```js
const plainDate = plainYearMonth.toPlainDate({ day: 1 });
const convertedPlainYearMonth = plainDate.withCalendar(localeCalendar).toPlainYearMonth();
convertedPlainYearMonth.toLocaleString();
```

More background reading in https://github.com/tc39/proposal-temporal/issues/262, https://github.com/tc39/proposal-temporal/issues/1221#issuecomment-755789111, and the [documentation](https://tc39.es/proposal-temporal/docs/plainmonthday.html#toLocaleString).

This is mandated by the spec in the §15.3, the additions to DateTimeFormat — for example, note the difference between [HandleDateTimeTemporalDate](https://tc39.es/proposal-temporal/#sec-temporal-handledatetimevaluetemporaldate) and [HandleDateTimeTemporalYearMonth](https://tc39.es/proposal-temporal/#sec-temporal-handledatetimevaluetemporaldate).

**Reply by mrcljx** [reply]
Thank you for this very exhaustive answer.

(I'll look into writing a lint rule to disallow `toLocaleString` on these objects without an explicit `calendar`.)
